pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -50,88 +51,115 @@ func init() {
+	daemonOverhead, err := s.getDaemonOverhead(ctx, constraints)
 	if err != nil {
-		return nil, fmt.Errorf("constructing nodeset, %w", err)
+		return nil, err
 	}
 
-	unschedulableCount := 0
-	for _, pod := range pods {
-		isScheduled := false
-		for _, node := range nodeSet.nodes {
-			if err := node.Add(pod); err == nil {
-				isScheduled = true
-				break
-			}
+	// We loop and retrying to schedule to unschedulable pods as long as we are making progress.  This solves a few
+	// issues including pods with affinity to another pod in the batch. We could topo-sort to solve this, but it wouldn't
+	// solve the problem of scheduling pods where a particular order is needed to prevent a max-skew violation. E.g. if we
+	// had 5xA pods and 5xB pods were they have a zonal topology spread, but A can only go in one zone and B in another.
+	// We need to schedule them alternating, A, B, A, B, .... and this solution also solves that as well.
+	var nodes []*Node
+	errors := map[*v1.Pod]error{}
+	q := NewQueue(pods...)
+	for {
+		pod, ok := q.Pop()
+		if !ok {
+			break
 		}
-		if !isScheduled {
-			n := NewNode(constraints, nodeSet.daemonResources, instanceTypes)
-			if err := n.Add(pod); err != nil {
-				unschedulableCount++
-				logging.FromContext(ctx).With("pod", client.ObjectKeyFromObject(pod)).Errorf("Scheduling pod, %s", err)
-			} else {
-				nodeSet.Add(n)
+		// Use existing node or create a node one
+		node := s.scheduleExisting(pod, nodes)
+		if node == nil {
+			node = NewNode(constraints, topology, daemonOverhead, instanceTypes)
+			// we created a new hostname, so ensure all topologies are aware
+			topology.Register(v1.LabelHostname, node.Hostname)
+
+			if errors[pod] = node.Add(pod); errors[pod] != nil {
+				relaxed := s.preferences.Relax(ctx, pod)
+				if relaxed {
+					// The pod has changed, so topology needs to be recomputed
+					if err := topology.Update(ctx, pod); err != nil {
+						logging.FromContext(ctx).With("pod", client.ObjectKeyFromObject(pod)).Errorf("updating topology, %s", err)
+					}
+				}
+				q.Push(pod, relaxed)
+				continue
 			}
+			nodes = append(nodes, node)
 		}
+
+		// Successfully scheduled the pod on a node, so record topology decision for future pods
+		topology.Record(pod, node.Constraints.Requirements)
 	}
 
-	if unschedulableCount != 0 {
-		logging.FromContext(ctx).Errorf("Failed to schedule %d pods", unschedulableCount)
+	// Any remaining pods have failed to schedule
+	for _, pod := range q.List() {
+		logging.FromContext(ctx).With("pod", client.ObjectKeyFromObject(pod)).Error(errors[pod])
 	}
-	return nodeSet.nodes, nil
+	return nodes, nil
 }
 
-func byPrice(instanceTypes []cloudprovider.InstanceType) func(i int, j int) bool {
-	return func(i, j int) bool {
-		return instanceTypes[i].Price() < instanceTypes[j].Price()
+func (s *Scheduler) scheduleExisting(pod *v1.Pod, nodes []*Node) *Node {
+	// Try nodes in ascending order of number of pods to more evenly distribute nodes, 100ms at 2000 nodes.
```

Comment:
It may vary depending on CPU, but that should be it.  We're just sorting s list of pointers by the pod count which is a constant operation to retrieve.